File: Algorithms/src/Sorts/QuickSort.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

package Sorts;

import java.util.List;

/**
 *
 * @author Ibrahim Seckin
 */
public class QuickSort {
    
    public void quickSort(List<Integer> numList){
        quickSort(numList , 0 , numList.size() - 1);
    }
    
    public void quickSort(int[] numArray){
        quickSort(numArray , 0 , numArray.length - 1);
    }
    
    public void quickSort(List<Integer> numList , int start, int end){
        
        int i = start;
        int k = end;
        int pivot = numList.get((start+end)/2);
            
            while(i <= k){
                
                while(numList.get(i) < pivot)
                    i++; //index of the first number that is greater from pivot in left part
                while(numList.get(k) > pivot)
                    k--;
                if(i <= k){
                    swap(numList,i,k);
                    i++;
                    k--;
                }
            }
            //swap(numList , start , k);
            if(start < k)
                quickSort(numList,start,k);
            if(i < end)
                quickSort(numList,i,end);
        }

    
    public void quickSort(int[] numArray , int start, int end){
        
        int i = start;
        int k = end;
        int pivot = numArray[(start+end)/2];
            
            while(i <= k){
                
                while(numArray[i] < pivot)
                    i++; //index of the first number that is greater from pivot in left part
                while(numArray[k] > pivot)
                    k--;
                if(i <= k){
                    swap(numArray,i,k);
                    i++;
                    k--;
                }
            }
            if(start < k)
                quickSort(numArray,start,k);
            if(i < end)
                quickSort(numArray,i,end);
        }

    
    public void swap(List<Integer> numList , int first , int second){
        
        int temp = numList.get(first);
        numList.set(first, numList.get(second));
        numList.set(second, numList.get(first));
    }
    
    public void swap(int[] numArray , int first , int second){
        
        int temp = numArray[first];
        numArray[first] = numArray[second];
        numArray[second] = temp;
    }
}

File: AlgorithmsC#/Sorting/QuickSort.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Sorting
{
    class QuickSort
    {

        #region listVersion
        public static void quickSort<T>(List<T> data)
        {
            quickSort(data, 0, data.Count-1);
        }

        private static int divide<T>(List<T> list, int start, int end)
        {
            int i = start;
            int k = end;
            T pivot = list[((start + end) / 2)];

                while (k >= i)
                {
                    while (Comparer<T>.Default.Compare(list[i], pivot) < 0)
                        i++; //index of the first number that is greater from pivot in left part
                    while (Comparer<T>.Default.Compare(list[k], pivot) > 0)
                        k--;

                    if (i <= k)
                    {
                        swap(list, i, k);
                        i++; k--;
                    }
                }
                //swap(numList, start, k);
                return i;     

        }


        private static void quickSort<T>(List<T> numList, int start, int end)
        {
            int index = divide(numList, start, end);
            if (start < index - 1) quickSort(numList, start, index - 1);
            if (index < end)       quickSort(numList, index, end);
        }

        private static void swap<T>(List<T> numList, int first, int second)
        {

            T temp = numList[first];
            numList[first] = numList[second];
            numList[second] = temp;
        }
           
        #endregion 


    #region arrayVersion

          public static void quickSort<T>(T[] data)
        {
            quickSort(data, 0, data.Length-1);
        }

        private static int divide<T>(T [] list, int start, int end)
        {
            int i = start;
            int k = end;
            T pivot = list[((start + end) / 2)];

                while (k >= i)
                {
                    while (Comparer<T>.Default.Compare(list[i], pivot) < 0)
                        i++; //index of the first number that is greater from pivot in left part
                    while (Comparer<T>.Default.Compare(list[k], pivot) > 0)
                        k--;

                    if (i <= k)
                    {
                        swap(list, i, k);
                        i++; k--;
                    }
                }
                //swap(numList, start, k);
                return i;     

        }


        private static void quickSort<T>(T[] numList, int start, int end)
        {
            int index = divide(numList, start, end);
            if (start < index - 1) quickSort(numList, start, index - 1);
            if (index < end)       quickSort(numList, index, end);
        }

        private static void swap<T>(T [] numList, int first, int second)
        {

            T temp = numList[first];
            numList[first] = numList[second];
            numList[second] = temp;
        }


    #endregion

    }

}
